tests/rptest/tests/log_compaction_test.py

```diff
@@ -0,0 +1,243 @@
+            try:
+                self._dispatch_random_partition_move(self.topic_spec.name, 0)
+                self._wait_for_move_in_progress(self.topic_spec.name,
+                                                0,
+                                                timeout=5)
+            except Exception as e:
+                reporter.exc = e
+
+        partition_move_thread = threading.Thread(
+            target=background_test_loop,
+            args=(PartitionMoveExceptionReporter, issue_partition_move),
+            kwargs={
+                'iterations': 5,
+                'sleep_sec': 1
+            })
+
+        producer = KgoVerifierProducer(
+            context=self.test_context,
+            redpanda=self.redpanda,
+            topic=self.topic_spec.name,
+            msg_size=self.msg_size,
+            msg_count=self.msg_count,
+            rate_limit_bps=self.rate_limit,
+            key_set_cardinality=self.key_set_cardinality,
+            tolerate_data_loss=False,
+            tombstone_probability=self.tombstone_probability,
+            validate_latest_values=True,
+            custom_node=self.preallocated_nodes)
+
+        consumer = KgoVerifierConsumerGroupConsumer(
+            self.test_context,
+            self.redpanda,
+            self.topic_spec.name,
+            self.msg_size,
+            readers=self.consumer_count,
+            compacted=True,
+            nodes=self.preallocated_nodes)
+
+        # Start partition movement thread
+        partition_move_thread.start()
+
+        # Start stress fibers
+        self._start_stress_fibers()
+
+        # Produce and wait
+        producer.start()
+        producer.wait_for_latest_value_map()
+        producer.wait(timeout_sec=60)
+
+        assert producer.produce_status.tombstones_produced > 0
+        assert producer.produce_status.bad_offsets == 0
+
+        # Consume and wait. clean=False to not accidentally remove latest value map.
+        consumer.start(clean=False)
+        consumer.wait(timeout_sec=60)
+
+        partition_move_thread.join()
+        self._stop_stress_fibers()
+
+        # Clean up
+        producer.stop()
+        consumer.stop()
+
+        if PartitionMoveExceptionReporter.exc is not None:
+            raise PartitionMoveExceptionReporter.exc
+
+        assert consumer.consumer_status.validator.tombstones_consumed > 0
+        assert consumer.consumer_status.validator.invalid_reads == 0
+
+    def part_two(self):
+        """
+        After several rounds of compaction, restart the brokers,
+        create a consumer, and assert that no tombstones are consumed.
+        """
+
+        # Restart each redpanda broker to force roll segments
+        self.redpanda.restart_nodes(self.redpanda.nodes)
+        self.redpanda.wait_for_membership(first_start=False)
+
+        # Sleep until the log has definitely been fully compacted.
+        timeout = self.extra_rp_conf['log_compaction_interval_ms'] / 1000 * 10
+        time.sleep(timeout)
```

Comment:
I think it doesn't guarantee that the log was compacted. It's better to read a metric to validate that compaction happened.

Reply:
I have added a number of metrics to the `storage::probe` that are now used in this test case, and will be of future use for compaction observability:

* `_segment_cleanly_compacted`, the number of segments that have been cleanly compacted (i.e, had their keys de-duplicated with all previous segments before them to the front of the log).
* `_segments_marked_tombstone_free`, the number of segments that have beenverified through the compaction process to be tombstone free.
* `_num_rounds_window_compaction`, the number of rounds of sliding window compaction that have been driven to completion.

I have removed this `sleep()` in place of a deterministic check that compaction has settled using some of these metrics.